x/staking/teststaking/handler.go

```diff
@@ -0,0 +1,64 @@
+package teststaking
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/staking"
+	"github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// HandlerT is a structure which wraps the staking handler
+// and provides methods useful in tests
+type HandlerT struct {
```

Comment:
I see what you're getting at here. Maybe there are cleaner ways to do this but this is one approach. I think the naming is confusing. How about just calling this something like `TestHelper`?

Reply:
`TestHelper` is too generic. I'm looking for a name which would represent something like a _handler service_. What I'm doing here is in fact a service for the module.

Reply:
The naming is confusing, though `TestHelper` does not help much (and among the two, `HandlerT` sounds better to me). I am definitely more concerned about the package's name - `teststaking` sounds imperative and makes very little sense. I'd suggest to rename it to just `testing`, e.g. the following looks much better from client developer's perspective IMHO: 

```golang
import stakingTesting "github.com/cosmos/cosmos-sdk/x/staking/testing"
```

Reply:
Re package rename. `testing` is a name of the standard (builtin) package, and shouldn't be used otherwise. 
This package is meant to use in tests, to provide helpers / mocks / stubs. I already went through it in past projects and it's good to group them under the package we will mock or provide those helpers.

Any other, shorter name will cause that we will have name conflicts and we will do same things as in other places - package aliases. This doesn't help because we will use it probably everywhere (to bring clarity) and our editors don't like it (because we need to add this packages + aliases manually).

Reply:
Re struct name, I renamed it to `Service`, because as I wrote in the first comment, it's meant to support module services. What do you think?

Reply:
How about `github.com/cosmos/cosmos-sdk/x/staking/testutil`? We already have some of those in the SDK.

Reply:
> Re package rename. `testing` is a name of the standard (builtin) package, and shouldn't be used otherwise.

Agreed

> Any other, shorter name will cause that we will have name conflicts and we will do same things as in other places - package aliases.

While this is true we don't do it in general, for instance we use `x/bank/types` not `x/bank/banktypes`. Going in that direction would be a larger refactor for the SDK that I'm not sure would support readability. (I do btw agree with the reasoning for `legacytx` but that's sort of an isolated case IMHO.)

I support `testutil.`

Reply:
> Re struct name, I renamed it to `Service`, because as I wrote in the first comment, it's meant to support module services. What do you think?

How about `HandlerHelper` or `ServiceHelper`? I just think `Helper` makes it clearer it's for tests.

Reply:
`test` is already in package name, so I don't think we need to use helper suffixes (neither in package name nor strict name). 
`testutil.Service` or `teststaking.Service` is short and clear (Go spirit)
Personally I prefer the latter.
The argument against the `testutil` is same as in my comments above: unnecessary generic. This package doesn't provide general testing utilities. If we will adopt that name then we will only increase the package aliases usage, which IMHO is not good (I explained it in my earlier comments)

Reply:
However if it's only me who doesn't like the `testutil` name, then ok, I will rename (ps, @alexanderbez  we were agreeing that package names like util, utils, common ... are not good, testutil is not better IMHO).